### PR TITLE
[improve][client] Add exception handle for client send error

### PR DESCRIPTION
**Pull request by congbobo184**
### Motivation

- producer send messages return error will close the current cnx
- if close the current cnx, all producers and consumers witch use this cnx will close and reconnect
(https://github.com/apache/pulsar/blob/5c6602cbb3660a696bf960f2847aac1a2ae037d2/pulsar-client/src/main/java/org/apache/pulsar/client/impl/ClientCnx.java#L323-L345)
- this will influence a lot of producers and consumers
- we only close this producer and reconnect enough, don't need to close this cnx

### Modifications

receive send_error, close current producer connection, then it will reconnect automatically

### Verifying this change

- [x] Make sure that the change passes the CI checks.

*(Please pick either of the following options)*

This change is a trivial rework / code cleanup without any test coverage.

*(or)*

This change is already covered by existing tests, such as *(please describe tests)*.

*(or)*

This change added tests and can be verified as follows:

*(example:)*
  - *Added integration tests for end-to-end deployment with large payloads (10MB)*
  - *Extended integration test for recovery after broker failure*

### Does this pull request potentially affect one of the following parts:

*If the box was checked, please highlight the changes*

- [ ] Dependencies (add or upgrade a dependency)
- [ ] The public API
- [ ] The schema
- [ ] The default values of configurations
- [ ] The threading model
- [ ] The binary protocol
- [ ] The REST endpoints
- [ ] The admin CLI options
- [ ] The metrics
- [ ] Anything that affects deployment

### Documentation

- [ ] `doc` 
- [ ] `doc-required` 
- [x] `doc-not-needed` 
- [ ] `doc-complete` 

### Matching PR in forked repository

PR in forked repository:

### CODE REVIEW

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ClientCnxTest.java [hunk 1]
```diff
@@ -25,16 +25,21 @@
 import java.util.concurrent.CountDownLatch;
 import java.util.concurrent.ExecutorService;
 import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+
```
Comment (BewareMyPower) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ClientCnxTest.java:29]:
```suggestion
import java.util.concurrent.TimeUnit;
```

**Review by BewareMyPower — COMMENTED** [review]
Fix code style

**Review by BewareMyPower — APPROVED** [review]

**Comment by BewareMyPower** [comment]
LGTM. But not sure why an extra blank line between the import clauses passes the checkstyle check.

**Review by dao-jun — APPROVED** [review]

**Review by HQebupt — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ClientCnxTest.java [hunk 2]
```diff
@@ -130,6 +134,50 @@ public void testClientVersion() throws Exception {
...
+                            && !cnxOne.equals(((ProducerImpl<?>) producerOne).getClientCnx()))
+                            || !cnxTwo.equals(((ProducerImpl<?>) producerTwo).getClientCnx()));
+            Assert.fail();
+        } catch (Throwable e) {
+            Assert.assertTrue(e instanceof ConditionTimeoutException);
+        }
+
+        // two producer use the same cnx
+        Assert.assertEquals(((ProducerImpl<?>) producerTwo).getClientCnx(),
+                ((ProducerImpl<?>) producerOne).getClientCnx());
```
Comment (nodece) [inline pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ClientCnxTest.java:171]: This check looks confusing, could you try `pollDelay`? which can meet your needs.

```suggestion
        // the cnx will not change
        Awaitility.await().pollDelay(3, TimeUnit.SECONDS)
                .until(() -> cnxOne.equals(((ProducerImpl<?>) producerOne).getClientCnx()) && cnxTwo.equals(
                        ((ProducerImpl<?>) producerTwo).getClientCnx()));
```

**Review by nodece — APPROVED** [review]

File: pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ClientCnxTest.java (hunk 2 above)
Comment (nodece) [reply pulsar-broker/src/test/java/org/apache/pulsar/client/impl/ClientCnxTest.java:171]: It has been merged.

**Pull request merged**